rascal2/widgets/delegates.py

```diff
@@ -1,40 +1,88 @@
 """Delegates for items in Qt tables."""
 
+from typing import Literal
+
 from PyQt6 import QtCore, QtGui, QtWidgets
 
+from rascal2.widgets.inputs import AdaptiveDoubleSpinBox, get_validated_input
+
 
-class EnumDelegate(QtWidgets.QStyledItemDelegate):
-    """Item delegate for Enums."""
+class ValidatedInputDelegate(QtWidgets.QStyledItemDelegate):
+    """Item delegate for validated inputs."""
 
-    def __init__(self, parent, enum):
+    def __init__(self, field_info, parent):
         super().__init__(parent)
-        self.enum = enum
+        self.table = parent
+        self.field_info = field_info
 
     def createEditor(self, parent, option, index):
-        combobox = QtWidgets.QComboBox(parent)
-        combobox.addItems(str(e.value) for e in self.enum)
-        return combobox
+        widget = get_validated_input(self.field_info, parent)
+        widget.set_data(index.data(QtCore.Qt.ItemDataRole.DisplayRole))
+
+        # without this signal, data isn't committed unless the user presses 'enter' on the cell
+        widget.edited_signal.connect(lambda: self.table.commitData(widget))
+
+        # fill in background as otherwise you can see the original View text underneath
+        widget.setAutoFillBackground(True)
+        widget.setBackgroundRole(QtGui.QPalette.ColorRole.Base)
+
+        return widget
 
-    def setEditorData(self, editor: QtWidgets.QCheckBox, index):
+    def setEditorData(self, editor: QtWidgets.QWidget, index):
         data = index.data(QtCore.Qt.ItemDataRole.DisplayRole)
-        editor.setCurrentText(data)
+        editor.set_data(data)
+
+    def setModelData(self, editor, model, index):
+        data = editor.get_data()
+        model.setData(index, data, QtCore.Qt.ItemDataRole.EditRole)
+
 
+class ValueSpinBoxDelegate(QtWidgets.QStyledItemDelegate):
+    """Item delegate for parameter values between a dynamic min and max.
 
-class BoolDelegate(QtWidgets.QStyledItemDelegate):
-    """Item delegate for bools."""
+    Parameters
+    ----------
+    field : Literal["min", "value", "max"]
+        The field of the parameter
 
-    def __init__(self, parent):
+    """
+
+    _error_style = "color: #E34234"
+
+    def __init__(self, field: Literal["min", "value", "max"], parent):
         super().__init__(parent)
-        self.parent = parent
+        self.table = parent
+        self.field = field
 
     def createEditor(self, parent, option, index):
-        checkbox = QtWidgets.QCheckBox(parent)
+        widget = AdaptiveDoubleSpinBox(parent)
+
+        # without this signal, data isn't committed unless the user presses 'enter' on the cell
+        widget.valueChanged.connect(lambda: self.table.commitData(widget))
```

Comment:
@StephenNneji one interesting thing going on: this line is producing a *lot* of stderr output from Qt (once per widget on startup then once per value change). according to the Qt source code, when the warning is produced then the function should do nothing:
https://github.com/qt/qtbase/blob/4678d96d874c21b69f4942e3bfdc5fc97128cf4e/src/widgets/itemviews/qabstractitemview.cpp#L3012-L3020
but when i remove the lines, the max/min changing in setModelData no longer works. not sure if you've seen anything like this or can advise on any alternatives?